### PR TITLE
*: add background goroutine to auto free os memory when process resident memory exceeds the  value

### PR DESCRIPTION
**Pull request by crazycs520**
Signed-off-by: crazycs520 <crazycs520@gmail.com>

### What problem does this PR solve?

Golang will not return the released memory to os, in some big transaction case, the `tidb-server` will be killed by os cause by OOM, by the `heapInuse` of `tidb-server` is not very much, just because golang doesn't return the idle memory to the os. 

### What is changed and how it works?

Add a server scope variable `tidb_auto_free_os_memory_threshold` to set the threshold of auto-free memory to the system.

The unit of `tidb_auto_free_os_memory_threshold` is byte, when the value of `tidb_auto_free_os_memory_threshold` more than `0`, and the process resident memory of `tidb-server` is more than `tidb_auto_free_os_memory_threshold`, `tidb-server` will auto do GC and return the idle memory to the os every 5 seconds.

Usage:

```sql
test> set @@tidb_auto_free_os_memory_threshold=1000;
Query OK, 0 rows affected
Time: 0.047s
test> select @@tidb_auto_free_os_memory_threshold;
+--------------------------------------+
| @@tidb_auto_free_os_memory_threshold |
+--------------------------------------+
| 1000                                 |
+--------------------------------------+
```

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Manual test 

Side effects

- Performance regression
    - Consumes more CPU

### Release note 

- Add background goroutine to auto-free os memory when process resident memory exceeds the  value

### CODE REVIEW

**Comment by crazycs520** [comment]
@lysu @tiancaiamao PTAL

**Comment by jackysp** [comment]
I remember @bobotu has some similar ideas to handle it.

**Comment by bobotu** [comment]
I cannot get the point of this PR. If there is no other process running on the same node, the reserved memory space shouldn't cause any problem. The reserved memory means the OS already allows the process to use such an amount of memory, why the process will be OOMed after that?

I don't think call `runtime.GC()` manually is a good idea, it may cause a very long STW pause if the sweep operation doesn't complished, or the `runtime.GC` itself will be halt in sweep operation before start any GC mark works.

In my point of view, OOM issues should be and can only be resolved in a JVM way
* Ask user to set a target heap size, such as 80% of physical RAM;
* GC will be triggered when (estimate GC time * average allocation rate >= free heap space), which means TiDB will not do any GC works until it will exceed the heap size limit, this can reduce GC overheap dramatically (in a naive sysbench oplt_read_only test, QPS can increased to 50k from 40k);

In this way, we can use the ununsed memory space to get better performance. If you don't agree with it, please tell me why the users paied moeny to get 128G RAM, and TiDB only use 10G of them and use large amount of CPU to do GC works and ask users to endure slow queries?

And the OOM problem will also go away, except the memory required by qurey is really exceeed the physical RAM limit. 

What we need to do is apply a small patch like https://github.com/bobotu/go/commit/8f8adda5d70b5e1531d86d7bc584bc96f92f3173 to Go and recompile TiDB with it.

File: util/systimemon/systime_mon.go [hunk 1]
```diff
@@ -41,3 +47,44 @@ func StartMonitor(now func() time.Time, systimeErrHandler func(), successCallbac
...
+		}
+
+		stat, err := p.Stat()
+		if err != nil {
+			logErrFn("get process stat failed", zap.Int("pid", pid), zap.Error(err))
+			continue
+		}
+		rss := stat.ResidentMemory()
+		if rss > threshold {
+			runtime.GC()
```
Comment (tiancaiamao) [inline util/systimemon/systime_mon.go:84]: FreeOSMemory will run a GC, so this is unnecessary.

File: tidb-server/main.go [hunk 2]
```diff
@@ -666,7 +666,7 @@ func setupMetrics() {
 		}
 	}
 	go systimemon.StartMonitor(time.Now, systimeErrHandler, sucessCallBack)
-
+	go systimemon.StartAutoFreeOSMemory()
```
Comment (tiancaiamao) [inline tidb-server/main.go:669]: This code only works on linux, how about adding

``` 
if runtime.GOOS == "Linux"
```

**Comment by crazycs520** [comment]
Better to use https://github.com/pingcap/tidb/pull/20253 instead use of this PR, so I close this PR.

**Pull request closed without merging**